[FLINK-21714][table-api]  Use TIMESTAMP_LTZ as return type for function PROCTIME()

leonardBang opened the pull request:
## What is the purpose of the change

* This pull request changes PROCTIME() function's return type to TIMESTAMP_LTZ as well as proctime attribute.

## Brief change log
    1. Change PROCTIME() function's return type to TIMESTAMP_LTZ
    2. support max/min function for TIMESTAMP_LTZ type
    3. Support TIMESTAMP_LTZ as proctime type for blink planner
    4. Support window operator base on TIMESTAMP_LTZ type, also support DaylightTime
    5. Support TIMESTAMP_LTZ as proctime type for temporal join operator
    6. Support TIMESTAMP_LTZ type as proctime for match recognize

## Verifying this change

* 1. add unit tests and IT tests to cover the function change and time attribute change

 
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (YES)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  NO, this PR is for pre review, will add docs later

abidaJustice commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecTemporalSort.java:
```diff
@@ -78,10 +79,15 @@ public StreamExecTemporalSort(
+
         throw new TableException(
                 String.format(
                         "Sort: Internal Error\n"
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

wuchong commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/TableToDataStreamITCase.scala:
```diff
@@ -256,6 +256,82 @@ final class TableToDataStreamITCase extends StreamingTestBase {
+    ds2.addSink(sink)
+    env.execute("DataStreamToTableBackDataStream")
+
+    // the result should like as following, we only check the result
```
ditto.

wuchong reviewed (commented):
I notices we use `String` to pass time zone at many places, actually, we should be able to use `TimeZone` type to avoid passing wrong value. Could you open another issue to improve this?

wuchong commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java:
```diff
@@ -215,6 +215,8 @@ public static DataFormatConverter getConverterForDataType(DataType originDataTyp
                 } else if (clazz == TimestampData.class) {
                     return new TimestampDataConverter(precisionOfLZTS);
+                } else if (clazz == Timestamp.class) {
+                    return new TimestampConverter(precisionOfLZTS);
```
This conversion will lost 8 hours if in Shanghai time zone. We may need a new converter for LTZ -> java.sql.Timestamp.

leonardBang commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -120,6 +120,10 @@ public static boolean isProctimeAttribute(LogicalType logicalType) {
         return logicalType.accept(TIMESTAMP_KIND_EXTRACTOR) == TimestampKind.PROCTIME;
     }
 
+    public static boolean isTimestampLtzType(LogicalType logicalType) {
```
will be used when we support rowtime on TIMESTAMP_LTZ

leonardBang commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/TimeWindowUtil.java:
```diff
@@ -0,0 +1,112 @@
+     * @return the mills which can describe the local timestamp string in given timezone.
+     */
+    public static long toUtcTimestampMills(long epochMills, TimeZone timeZone) {
+        if (timeZone.toZoneId().equals(UTC_ZONE_ID)) {
+            return epochMills;
+        }
+        LocalDateTime localDateTime =
+                LocalDateTime.ofInstant(Instant.ofEpochMilli(epochMills), timeZone.toZoneId());
+        return localDateTime.atZone(UTC_ZONE_ID).toInstant().toEpochMilli();
```
I think we can not use  `TimeZone` to compare

leonardBang commented:
@flinkbot run azure

leonardBang commented:
@flinkbot run azure

leonardBang commented:
@wuchong @KurtYoung the build passed，please help merge.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java:
```diff
@@ -248,11 +252,11 @@ private void validateTimeColumn(String columnName, List<Column> columns) {
 
     private void validateWatermarkExpression(LogicalType watermarkType) {
-        if (!hasRoot(watermarkType, LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+        if (!hasFamily(watermarkType, LogicalTypeFamily.TIMESTAMP)
```
with this you also allow timestamp's with time zone, we should be more precise here

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java:
```diff
@@ -248,11 +252,11 @@ private void validateTimeColumn(String columnName, List<Column> columns) {
             throw new ValidationException(
                     "Invalid data type of expression for watermark definition. "
-                            + "The field must be of type TIMESTAMP(3) WITHOUT TIME ZONE.");
+                            + "The field must be of type TIMESTAMP(3) WITHOUT TIME ZONE or TIMESTAMP_LTZ(3) WITHOUT TIME ZONE.");
```
remove last `WITHOUT TIME ZONE`

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java:
```diff
@@ -270,13 +274,29 @@ private Column adjustRowtimeAttribute(List<WatermarkSpec> watermarkSpecs, Column
+                default:
+                    throw new ValidationException(
+                            "Invalid data type of expression for rowtime definition. "
+                                    + "The field must be of type TIMESTAMP(3) WITHOUT TIME ZONE or TIMESTAMP_LTZ(3) WITHOUT TIME ZONE.");
```
remove last `WITHOUT TIME ZONE`

twalthr reviewed (changes requested):
This is a massive PR @leonardBang. I would also like to take another look. My first scan revealed already some bugs. Could we wait another day before merging this? I think after feature freeze is a good time to get such a thing in.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/LocalZonedTimestampType.java:
```diff
@@ -72,7 +72,8 @@
                     Long.class.getName(),
-                    TimestampData.class.getName());
+                    TimestampData.class.getName(),
+                    java.sql.Timestamp.class.getName());
```
Are we still planning to support `LocalDateTime` as well? Or is `java.sql.Timestamp` enough for us?  Supporting `LocalDateTime` would mean that we need access to configuration in converters, correct?

leonardBang replied:
yes, we think `java.sql.Timestamp` is enough, Supporting `LocalDateTime` is an impossible mission which means we need pass configuration to converters, but some place we're hard to get the configuration eg. format , serializers

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -120,6 +120,25 @@ public static boolean isProctimeAttribute(LogicalType logicalType) {
         return logicalType.accept(TIMESTAMP_KIND_EXTRACTOR) == TimestampKind.PROCTIME;
     }
 
+    public static boolean supportedTimeAttributeType(LogicalType logicalType) {
```
nit: it would be nice to start methods with verbs e.g. `isTimeAttributeType` or `isWatermarkType` which is also shorter

wuchong replied:
What about `canBeTimeAttributeType`?

twalthr replied:
+1

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/LocalZonedTimestampType.java:
```diff
@@ -81,7 +82,8 @@
                     long.class.getName(),
-                    TimestampData.class.getName());
+                    TimestampData.class.getName(),
+                    java.sql.Timestamp.class.getName());
```
Please also update the documentation table.

twalthr commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGroupWindowAggregate.java:
```diff
@@ -207,6 +208,10 @@ public StreamExecGroupWindowAggregate(
         }
 
+        final String shiftTimeZone =
+                TimeWindowUtil.getShiftTimeZone(
```
+1 to Jark's question

twalthr commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecTemporalSort.java:
```diff
@@ -104,10 +105,15 @@ public StreamExecTemporalSort(
+        }
+        if (timeType instanceof LocalZonedTimestampType) {
+            LocalZonedTimestampType keyType = (LocalZonedTimestampType) timeType;
+            if (keyType.getKind() == TimestampKind.PROCTIME) {
```
If `PROCTIME` will always be `LocalZonedTimestampType` from now on, shall we forbid this kind in the constructor of `ZonedTimestampType` and `TimestampType`? Would it make our life easier?

twalthr commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -602,10 +612,17 @@ class RexTimeIndicatorMaterializer(
-    .getTypeFactory
-    .asInstanceOf[FlinkTypeFactory]
-    .createFieldTypeFromLogicalType(new TimestampType(isNullable, 3))
+  private def timestamp(isNullable: Boolean, isTimestampLtz: Boolean = false): RelDataType = {
```
reduce 3 code duplication and refactor this to a global method in this file that takes `RexBuilder`?

twalthr commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java:
```diff
@@ -900,6 +902,36 @@ Timestamp toExternalImpl(RowData row, int column) {
     }
 
+    /** Converter for timestamp which doesn't consider the time zone. */
+    public static final class TimestampLtzConverter
```
Please also update the non-legacy `org.apache.flink.table.data.conversion.DataStructureConverters`

twalthr commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/TimeWindowUtil.java:
```diff
@@ -0,0 +1,134 @@
+import java.util.TimeZone;
+
+/** Time util to deals window start and end in different timezone. */
+public class TimeWindowUtil {
```
mark as `@Internal`

twalthr reviewed (changes requested):
Thanks for the PR @leonardBang. I added some comments. In general this PR is great work. But please try to split such PRs into smaller commits in the future. E.g. the conversion class for `LocalZonedTimestamp` could have been an easier PR on its own. Also the guidelines are very clear about this (even in bold letters):
https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html

leonardBang commented:
@twalthr Thanks very much for your comments and advice that I should have opened another PR for some subtasks, I'll keep in mind later in next work.
@twalthr @wuchong I should have addressed your comments and add more tests, could you have a more look? thanks

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/agg/AggsHandlerCodeGenerator.scala:
```diff
@@ -1082,6 +1100,17 @@ class AggsHandlerCodeGenerator(
+     } else {
+       s"""
+          |$TIME_WINDOW_UTIL.toEpochMills(
+          |$itemExpr, java.time.ZoneId.of("${shiftTimeZone.getId}"))
```
nit: Personally, I suggest to serialize `ZoneId` as a member field to avoid construct the `ZoneId` per record.

wuchong commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/SqlGroupedWindowFunction.java:
```diff
@@ -0,0 +1,191 @@
+ * and its auxiliary functions are {@code HOP_START} and {@code HOP_END}. Here they are used in a
+ * streaming query:
+ *
+ * <p>Note: we copied the implementation from Calcite's {@link
```
Add Calcite JIRA id.

wuchong commented on flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/data/DataFormatConvertersTest.java:
```diff
@@ -85,7 +85,7 @@
                 LocalTimeTypeInfo.LOCAL_TIME,
                 LocalTimeTypeInfo.LOCAL_DATE_TIME,
-                StringDataTypeInfo.INSTANCE
+                StringDataTypeInfo.INSTANCE,
```
remove end `,`?

wuchong reviewed (approved):
Thanks for the updating. The pull request looks good to me overall, I only left 3 minor comments.

wuchong commented:
Merging...
